internal/k8s/reconciler/listener.go

```diff
@@ -119,12 +140,109 @@ func (l *K8sListener) validateTLS(ctx context.Context) error {
 		}
 		l.status.ResolvedRefs.InvalidCertificateRef = certificateErr
 	} else {
-		l.certificates = []string{resource}
+		if l.tlsParams != nil {
+			l.tlsParams.certificates = []string{resource}
+		} else {
+			l.tlsParams = &tlsParams{
+				certificates: []string{resource},
+			}
+		}
+	}
+
+	if l.listener.TLS.Options != nil {
+		tlsMinVersion := l.listener.TLS.Options[tlsMinVersionAnnotationKey]
+		tlsMaxVersion := l.listener.TLS.Options[tlsMaxVersionAnnotationKey]
+		tlsCipherSuitesStr := l.listener.TLS.Options[tlsCipherSuitesAnnotationKey]
+
+		if tlsMinVersion != "" {
+			if _, ok := supportedTlsVersions[string(tlsMinVersion)]; !ok {
+				l.status.Ready.Invalid = errors.New("unrecognized TLS min version")
+				return nil
+			}
+
+			if tlsCipherSuitesStr != "" {
+				if _, ok := tlsVersionsWithConfigurableCipherSuites[string(tlsMinVersion)]; !ok {
+					l.status.Ready.Invalid = errors.New("configuring TLS cipher suites is only supported for TLS 1.2 and earlier")
+					return nil
+				}
+			}
+
+			if l.tlsParams != nil {
+				l.tlsParams.minVersion = string(tlsMinVersion)
+			} else {
+				l.tlsParams = &tlsParams{
+					minVersion: string(tlsMinVersion),
+				}
+			}
+		}
+
+		if tlsMaxVersion != "" {
+			if _, ok := supportedTlsVersions[string(tlsMaxVersion)]; !ok {
+				l.status.Ready.Invalid = errors.New("unrecognized TLS max version")
+				return nil
+			}
+
+			if l.tlsParams != nil {
+				l.tlsParams.maxVersion = string(tlsMaxVersion)
+			} else {
+				l.tlsParams = &tlsParams{
+					maxVersion: string(tlsMaxVersion),
+				}
+			}
+		}
+
+		if tlsCipherSuitesStr != "" {
+			// split comma delimited string into string array and trim whitespace
+			tlsCipherSuitesUntrimmed := strings.Split(string(tlsCipherSuitesStr), ",")
+			tlsCipherSuites := tlsCipherSuitesUntrimmed[:0]
+			for _, c := range tlsCipherSuitesUntrimmed {
+				tlsCipherSuites = append(tlsCipherSuites, strings.TrimSpace(c))
+			}
+
+			// validate each cipher suite in array
+			for _, c := range tlsCipherSuites {
+				if _, ok := supportedTlsCipherSuites[c]; !ok {
+					l.status.Ready.Invalid = fmt.Errorf("unrecognized or unsupported TLS cipher suite: %s", c)
+					return nil
+				}
+			}
+
+			// set cipher suites on listener TLSParams
+			if l.tlsParams != nil {
+				l.tlsParams.cipherSuites = tlsCipherSuites
+			} else {
+				l.tlsParams = &tlsParams{
+					cipherSuites: tlsCipherSuites,
+				}
+			}
```

Comment:
Checking this for every possible option is a bit tedious, should `K8sListener.tlsParams` maybe not be a pointer?

Reply:
Changed this to not be a pointer when switching to `core.TLSParams`